// @expo-google-fonts/dm-sans/index.js
// This file is generated by @expo-google-fonts/generator
// If you want to modify it, go to packages/generator/templates

export * from './useFonts';
export { default as __metadata__ } from './metadata.json';
export const DMSans_100Thin = require('./100Thin/DMSans_100Thin.ttf');
export const DMSans_200ExtraLight = require('./200ExtraLight/DMSans_200ExtraLight.ttf');
export const DMSans_300Light = require('./300Light/DMSans_300Light.ttf');
export const DMSans_400Regular = require('./400Regular/DMSans_400Regular.ttf');
export const DMSans_500Medium = require('./500Medium/DMSans_500Medium.ttf');
export const DMSans_600SemiBold = require('./600SemiBold/DMSans_600SemiBold.ttf');
export const DMSans_700Bold = require('./700Bold/DMSans_700Bold.ttf');
export const DMSans_800ExtraBold = require('./800ExtraBold/DMSans_800ExtraBold.ttf');
export const DMSans_900Black = require('./900Black/DMSans_900Black.ttf');
export const DMSans_100Thin_Italic = require('./100Thin_Italic/DMSans_100Thin_Italic.ttf');
export const DMSans_200ExtraLight_Italic = require('./200ExtraLight_Italic/DMSans_200ExtraLight_Italic.ttf');
export const DMSans_300Light_Italic = require('./300Light_Italic/DMSans_300Light_Italic.ttf');
export const DMSans_400Regular_Italic = require('./400Regular_Italic/DMSans_400Regular_Italic.ttf');
export const DMSans_500Medium_Italic = require('./500Medium_Italic/DMSans_500Medium_Italic.ttf');
export const DMSans_600SemiBold_Italic = require('./600SemiBold_Italic/DMSans_600SemiBold_Italic.ttf');
export const DMSans_700Bold_Italic = require('./700Bold_Italic/DMSans_700Bold_Italic.ttf');
export const DMSans_800ExtraBold_Italic = require('./800ExtraBold_Italic/DMSans_800ExtraBold_Italic.ttf');
export const DMSans_900Black_Italic = require('./900Black_Italic/DMSans_900Black_Italic.ttf');
